### ISSUE TITLE
closing webview windows causes crashes on some machines

### ISSUE THREAD

**Issue by dae** [ankitects/anki#1879]
Appears to be an issue on both [Windows](https://forums.ankiweb.net/t/mac-app-crashes-everytime-when-i-close-a-window/18792) and [Mac](https://forums.ankiweb.net/t/clicking-ok-in-about-anki-window-on-mac-causes-anki-to-crash/19993). I've tried two different Windows 10 machines, a Windows 11 machine, an Intel Mac, and an ARM Mac, but have not been able to reproduce it on any of those devices.

Presumably it's occurring in deferred_delete_and_garbage_collection() or the subsequent timer firing. If a dev is able to reproduce the issue on their machine, they could try experimenting with manually destroying specific properties such as _page/assigning a new blank page to see if that makes any difference.

The first linked report above says they're only seeing this in 6.3.0, though there have been reports in previous Anki versions of [similar issues](https://forums.ankiweb.net/t/mac-app-crashes-everytime-when-i-close-a-window/18792) that I could not reproduce either. So it's not clear whether this is a new bug introduced in 6.3.0, or whether it's a race condition that has just started affecting other users due to changes in timing.

Qt are only going to release 6.2.5 to commercial customers, which means we're stuck on 6.3.x now if we want the pasting/IME fixes coming in 6.3.1.

**Comment by mat1696** [comment]
I'm seeing this issue too on two different Macs.

**Comment by SamuelSwartzberg** [comment]
It appears I'm having the same issue mentioned here / in #1883, or a similar one, with no addons enabled, on `Version ⁨2.1.52 (ab1c2395)⁩; Python 3.9.7 Qt 6.3.0 PyQt 6.3.0`, on an M1 mac @ 12.0.1. However, I'm less confident in exactly what is causing it. I can reproduce it by opening the about/card editing menu/browse screen and closing it, both with the keyboard and the mouse. The race condition theory seems plausible insofar as that triggering it is more common if one closes the window quickly after opening it, or very quickly after using the HTML editing feature, for example. This might be unrelated, but after such a crash I also have to quit Anki e.g. via the Activity Monitor, as there is still a process from Anki that prevents reopening it. I also on a whim tried disabling my japanese input, which did nothing to help the problem.

**Comment by dae** [comment]
Maybe updating to the latest macOS will help?

**Comment by SamuelSwartzberg** [comment]
I mean, I should, you're right. (I got bricked by updating recently on a different device, so I'm hesitant, but it is probably necessary.) So I will do that and report back! But it does seem like it's not just me having some version of this problem.

**Comment by SamuelSwartzberg** [comment]
@dae Updated to 12.4, still persists.

**Comment by mat1696** [comment]
> Maybe updating to the latest macOS will help?

Unfortunately it doesn’t help: I have one Mac on 12.4 release and the other on 12.5 beta, and exactly the same issue as @SamuelSwartzberg . And I have to manually quit Anki process from Activity Monitor as well…

**Comment by FrostyGelato** [comment]
For me, the crash occurs when I click blue qt buttons but not when I click the red x button. In the About Anki window, for instance, the app crashes if I click the blue "OK" button in the bottom right corner but Anki does not crash if I click the red x button (the one that comes with all mac apps) in the top-left corner.

Not sure how useful this information is.

**Comment by dae** [comment]
I tried this on 2 Windows machines, a Linux machine and a Mac, and still wasn't able to reproduce the issue - opening 2.1.52 or 53rc1, opening the About screen then clicking on OK or pressing space to activate it didn't trigger a crash. I repeated the open/close steps multiple times, varying the amount of time between closing and reopening, but didn't experience a single crash. I wish I knew what the trick/environment was to trigger it, because then I'd have a hope of trying to find a workaround.

@RumovZ @hgiesel @abdnh @kleinerpirat @hikaru-y does it happen for any of you?

**Comment by SamuelSwartzberg** [comment]
@FrostyGelato Are you sure that's not down to timing? I can also make it crash with the red close button, but it's harder to click quickly from wherever your cursor normally is. But when I pre-place my cursor, I can reliably make it crash with the red exit button as well.  

On a different note, I opened the executable `anki` from within the package from the CLI to see the logs, and this Is what I got just before crashing:

```
Starting main loop...
2022-05-24 04:26:59.304 anki[9906:146661] TSM AdjustCapsLockLEDForKeyTransitionHandling - _ISSetPhysicalKeyboardCapsLockLED Inhibit
ignored late bridge cmd domDone
Segmentation fault: 11
```

**Comment by dae** [comment]
Could you guys also confirm you see the same issue in 2.1.53rc1?

https://forums.ankiweb.net/t/anki-crashes-on-mac/19247/16 

https://betas.ankiweb.net/#packaged-builds

**Comment by abdnh** [comment]
I'm having daily crashes on 2.1.52. I just had a crash after about 25 minutes of reviewing. It appears the issue always occurs after closing windows (happened for me after closing the editor in the last few days). I have not figured out how to reliably trigger it, but I feel it's more likely to occur with longer review sessions. Will test today on the latest main and 2.1.53rc1.

**Comment by SamuelSwartzberg** [comment]
> Could you guys also confirm you see the same issue in 2.1.53rc1?

Just tested, still occurs on 2.1.53rc1 for me.
On the reproducibility front, I can reproduce it 100% of the time if I open the note editor on a note and press esc very quickly, almost 100% of the time if I do it on the browser, and only like 15% of the time for the about menu with the same procedure for me.

**Comment by oakkitten** [comment]
It's interesting that when I am able to reproduce the crash, I can reproduce it repeatedly very reliably. But then sometimes changes and poof! I can no longer reproduce it no matter what I try. Windows 10 here.

**Comment by FrostyGelato** [comment]
The problem still occurs with 2.1.53rc1 qt 6 for mac. 

@SamuelSwartzberg It might be down to timing. After you pointed it out, I realised that Ok button in About Anki window does not crash the app if I wait for a few seconds before clicking it.

**Comment by RumovZ** [comment]
No crashes here.

**Comment by Neosaurus** [comment]
I also have this issue - I also have to force close Anki via activity monitor to reopen it again. I can consistently reproduce it via typing in a text field in the Add Card GUI, then pressing Esc and confirming closing the window. It does this regardless of timing. If there is no text in the field (i.e I don't need to confirm I want to close the window) the crash does not occur. I do also often experience random crashes which also require an activity monitor restart, most commonly in the Add Card GUI or Browser. I do have a fair number of apps enabled in accessibility settings though, and weeding out which one is causative something I unfortunately don't have time for right now, but they are:

- Alfred
- AltTab
- Anki Intel 2.1.52 (though I use the apple silicon version now on .52, which does not have access)
- Bartender
- BeardedSpice
- Google Drive
- JoyMapperSilicon
- KeyboardMaestro + KeyboardMaestro Engine
- Keyboard Clean Tool
- Logi Options
- Middle
- Monitor Control
- Rectangle Pro
- SoundSource
- Terminal

I am using a 16gb, 8 core 14" MBP on MacOS 12.3.1

**Comment by mat1696** [comment]
> I also have this issue - I also have to force close Anki via activity monitor to reopen it again. I can consistently reproduce it via typing in a text field in the Add Card GUI, then pressing Esc and confirming closing the window. It does this regardless of timing. If there is no text in the field (i.e I don't need to confirm I want to close the window) the crash does not occur. I do also often experience random crashes which also require an activity monitor restart, most commonly in the Add Card GUI or Browser. These all occur regardless of addons being enabled or not. I do have a fair number of apps enabled in accessibility settings though, and weeding out which one is causative is something I unfortunately don't have time for right now, but they are:
> 
> * Alfred
> * AltTab
> * Anki Intel 2.1.52 (though I use the apple silicon version now on .52, which does not have access)
> * Bartender
> * BeardedSpice
> * Google Drive
> * JoyMapperSilicon
> * KeyboardMaestro + KeyboardMaestro Engine
> * Keyboard Clean Tool
> * Logi Options
> * Middle
> * Monitor Control
> * Rectangle Pro
> * SoundSource
> * Terminal
> 
> I am using a 16gb, 8 core 14" MBP on MacOS 12.3.1

Same as you, but I don’t have any app in Accessibility, except Terminal… So don’t think it is related.

**Comment by abdnh** [comment]
Probably not very useful information, but Event Viewer on Windows 10 shows the following log:
![image](https://user-images.githubusercontent.com/41397710/170122465-a45e45c4-22ca-4109-8804-4b6d090af6c2.png)

```
Faulting application name: anki.exe, version: 0.0.0.0, time stamp: 0x6279dde1
Faulting module name: ucrtbase.dll, version: 10.0.19041.789, time stamp: 0x2bd748bf
Exception code: 0xc0000409
Fault offset: 0x000000000007286e
Faulting process id: 0x1e60
Faulting application start time: 0x01d86f0d68d73f49
Faulting application path: C:\Program Files\Anki\anki.exe
Faulting module path: C:\WINDOWS\System32\ucrtbase.dll
Report Id: 0010a719-b4e7-4566-8f04-aeab6dda7c5c
Faulting package full name: 
Faulting package-relative application ID: 
```

About the 0xc0000409 error code (which is not very informative, apparently):
https://stackoverflow.com/questions/23409809/windows-7-exception-code-0xc0000409
https://devblogs.microsoft.com/oldnewthing/20190108-00/?p=100655

I've not managed to reproduce it on the latest main so far.

**Comment by oakkitten** [comment]
Can't try the beta at the moment, but running this code from the debug console crashes Anki 2.1.52 100% of the time, and doesn't crash Anki 2.1.51, although I definitely saw similar crashes before:
```python
aqt.dialogs.open("AddCards", mw).close()
```
When I shove in a bit of coarse tracing, it always stops at the same line:

    $ ./anki-console.bat
    Anki starting...
    ...
    sys why='return', arg='null', frame=<frame at 0x000001FDE116E600, file 'json', line 231, code dumps>
    sys why='return', arg='null', frame=<frame at 0x000001FDDF47C440, file 'aqt.webview', line 42, code cmd>
    Press any key to continue . . .

Which is the return line of this method:

```python
@pyqtSlot(str, result=str)  # type: ignore
def cmd(self, str: str) -> Any:
    return json.dumps(self.onCmd(str))
```

It seems that `pyqtSlot` has something to do with threads? Oh well I have no idea. All trace messages come from a single thread.

Using Anki 2.1.52 on Windows 10. The core for the super duper tracer:

```python
import sys
import threading
from functools import partial

def trace(frame, why, arg, prefix):
    sys.__stdout__.write(f"{prefix} {why=}, {arg=}, {frame=}\n")
    sys.__stdout__.flush()
    from functools import partial
    return partial(sys.trace, prefix=prefix)

sys.trace = trace

dialog = aqt.dialogs.open("AddCards", mw)

sys.settrace(partial(trace, prefix="sys"))
threading.settrace(partial(trace, prefix="thr"))

dialog.close()
```

P.S. I sure wish debug console played nice with globals

**Comment by dae** [comment]
The open/close doesn't trigger it for me, but at least we have a developer with a reproducible test case now. Appreciate you looking into it.

Webviews use Qt's 'webchannel' functionality for sending messages back to Python code. This likely uses Qt's cross-thread 'queued' signal/slot mechanism. If we assume that the trace is accurately identifying the crash site, it seems to imply that the webview has called pycmd(), the message has been queued up to be delivered to the cmd() slot on the UI thread, and the message ends up being processed after the webview has been partially/fully deallocated. That shouldn't happen though, as deleteLater() is supposed to clear out pending events first.

Removing `deleteLater()` might fix the add cards screen, but it will cause other screens like the stats window and about screen to leak - the webengine process won't end up getting cleaned up when they are closed.

If you add back deleteLater() to patch(), and add any of the following changes to the top (before deleteLater()), do they help?

```
        obj._page = None
```

```        
        obj._onBridgeCmd = lambda x: None
```

```
        obj._page._bridge.cmd = lambda x: None
```

```
        obj._page._channel = QWebChannel()
        obj._page.setWebChannel(obj._page._channel)
```

**Comment by oakkitten** [comment]
Sorry, none of these help. To be clear, I was running stuff like this:
```python
original_method = aqt.webview.AnkiWebView.cleanup

def patch(self, original_method=original_method):
    self._page = None
    original_method(self)

aqt.webview.AnkiWebView.cleanup = patch
aqt.dialogs.open("AddCards", mw).close()
```
By the way, when the crash happens, the dialog disappears, and the whole app hangs for about a second. Sometimes less, sometimes a bit more, but it doesn't just crash at once.

I tried calling `deleteLater` a bit delayed. This works like 95% of the time:

```python
def patch(obj, mw=mw):
  mw.progress.single_shot(500, obj.deleteLater, False)
  mw.progress.single_shot(1000, mw.garbage_collect_now, False)

mw.deferred_delete_and_garbage_collect = patch
aqt.dialogs.open("AddCards", mw).close()
```
Changing 500 to 400 makes it crash about 50% of the time, and changing it to 300 makes it crash every time. Note that my computer is rather slow (x220, i5-2520M)

**Comment by oakkitten** [comment]
Ok this will sound like a joke, but you might want to try putting your computer to sleep briefly, and then try to reproduce the issue.

I restarted the computer and the problem *poof* went away. I tried all things that I usually do and finally had the idea to put the computer to sleep, just for a few seconds. Anki that was launched before sleep was fine after sleep as well. But when I launched a new copy of Anki, the crash returned and started happening reliably.

![image](https://user-images.githubusercontent.com/1710718/170150350-577c540f-417b-47fe-a703-a31a248ac5be.png)

**Comment by TRIAEIOU** [comment]
> I also have this issue - I also have to force close Anki via activity monitor to reopen it again. I can consistently reproduce it via typing in a text field in the Add Card GUI, then pressing Esc and confirming closing the window. It does this regardless of timing. If there is no text in the field (i.e I don't need to confirm I want to close the window) the crash does not occur. I do also often experience random crashes which also require an activity monitor restart, most commonly in the Add Card GUI or Browser. These all occur regardless of addons being enabled or not. I do have a fair number of apps enabled in accessibility settings though, and weeding out which one is causative is something I unfortunately don't have time for right now, but they are:
> 
> * Alfred
> * AltTab
> * Anki Intel 2.1.52 (though I use the apple silicon version now on .52, which does not have access)
> * Bartender
> * BeardedSpice
> * Google Drive
> * JoyMapperSilicon
> * KeyboardMaestro + KeyboardMaestro Engine
> * Keyboard Clean Tool
> * Logi Options
> * Middle
> * Monitor Control
> * Rectangle Pro
> * SoundSource
> * Terminal
> 
> I am using a 16gb, 8 core 14" MBP on MacOS 12.3.1

If it helps, I have the same problem on a i7 16GB RAM laptop running Windows 11: No problem closing windows (browser, editor) with the buttons but using escape crashes Anki completely with no console output (last thing seen is "ignored late bridge cmd blur:0:0:") - this is regardless of disabling all addons ("shift start") or not.

**Comment by BlueGreenMagick** [comment]
I can reliably reproduce this error on my m1 macbook, running macOS 12.3.1. Using v2.1.53rc1, qt6, both intel and arm versions. I can also reproduce it on latest master (9b07299).

> `aqt.dialogs.open("AddCards", mw).close()`
This code provided by @oakkitten reliably triggers the crash as well.

Commenting out the line `obj.deleteLater()` from `main.deferred_delete_and_garbage_collect()` does indeed fix the crash.

I can provide the crash log provided by Apple if it may help.

**Comment by dae** [comment]
I have finally managed to trigger this on a Windows 11 machine, and am digging into it.

**Comment by dae** [comment]
I'll get a new beta build out tomorrow for people to test with, so we can see if that has addressed the whole issue or just our reproducible test case.

**Comment by dae** [comment]
Please let us know how rc2 goes: https://betas.ankiweb.net/anki2.1.53.html#release-candidate-2

**Comment by SamuelSwartzberg** [comment]
Preliminarily, I don't seem to be able to reproduce crashes anymore, neither are they happening during normal usage. So, carefully optimistic!

**Comment by FrostyGelato** [comment]
Rc2 does not experience crashes when closing About Anki window or closing the edit card window.

**Comment by dae** [comment]
2.1.53 is out today with this fix (it's identical to 2.1.53rc2). For anyone still experiencing crashes after upgrading, please post on the forums so we can triage there.
